product_default_image/models/product_template.py

```diff
@@ -0,0 +1,299 @@
+
+        if target in ['category', 'global_category'] and vals.get('categ_id'):
+            categ = self.env['product.category'].browse(vals['categ_id'])
+            vals.update({
+                'image': categ.image,
+                'img_type': 'default_category',
+            })
+
+        if target in ['global', 'global_category'] and not vals.get('image'):
+            vals.update({
+                'image': company.product_image,
+                'img_type': 'default_global',
+            })
+
+        tools.image_resize_images(vals)
+        return vals
+
+    @api.model
+    def search_change_images(self, from_types, to_type, to_img_bg=None,
+                             add_domain=None, in_cache=False):
+
+        templates = self.search_by_img_types(
+            from_types=from_types,
+            add_domain=add_domain,
+        )
+
+        if not templates:
+            return
+
+        img_args = {
+            'to_type': to_type,
+            'in_cache': in_cache,
+        }
+        if to_img_bg:
+            img_args['to_img_bg'] = to_img_bg
+
+        return templates.change_template_image(**img_args)
+
+    @api.model
+    def search_by_img_types(self, from_types, add_domain=None):
+        """ Constructs domain and searches using from_types, add_domain
+
+            Args:
+                from_types:
+                    List of selections from img_type field to search
+                    products by.
+
+                add_domain:
+                    If you wish to add domain items to the method's default
+                    ones, use this parameter to do so. Must be a list.
+
+            Example:
+                .. code-block:: python
+
+                templates = self.env['product.template'].search_by_img_types(
+                    from_types=['default_category', 'no_image'],
+                    add_domain=[('categ_id', '=', record.id)],
+                )
+
+            Returns:
+                The above example will return a recordset using a domain of:
+                    .. code-block:: python
+
+                    [
+                        ('company_id', '=', some_id_int),
+                        ('auto_change_image', '=', True),
+                        ('img_type', 'in', ['default_category',
+                                              'no_image']),
+                        ('categ_id', '=', some_id_int),
+                    ]
+
+        """
+        if not add_domain:
+            add_domain = []
+
+        if not isinstance(from_types, list):
+            raise TypeError('from_types argument must be a list')
+
+        if add_domain and not isinstance(add_domain, list):
+            raise TypeError('add_domain argument must be a list')
+
+        domain = [
+            ('company_id', '=', self.env.user.company_id.id),
+            ('auto_change_image', '=', True),
+            ('img_type', 'in', from_types),
+        ]
+        domain += add_domain
```

Comment:
The domain part should be generated in a separate method in order to allow for modularity. On this note, the `add_domain` would likely become irrelevant - which I think you added in order to achieve the desired modularity?